odoo/addons/base/models/res_partner.py

```diff
@@ -517,7 +517,18 @@ def write(self, vals):
             self.invalidate_cache(['user_ids'], self._ids)
             for partner in self:
                 if partner.active and partner.user_ids:
-                    raise ValidationError(_('You cannot archive a contact linked to an internal user.'))
+                    user = self.env['res.users'].sudo(False).browse(partner.user_ids.id)
+                    try:
+                        user.check_access_rights('write')
+                        user.check_access_rule('write')
+                    except AccessError:
+                        raise ValidationError(_('You cannot archive a contact linked to an active user. '
```

Comment:
As I said in my previous comment, I believe the `check_access_rights` on `res.users` is enough for your scenario, checking on the record is overkill IMHO. We are talking about `res.partner` on which you have the right to write but not on `res.users`. The scenario where you can write on `res.users` model but not on an individual record is most likely a mutli-company issue and you should not be writing on the `res.partner` record in the first place. The worst case scenario without making the `check_access_rule` is that you say "You first need to archive its associated user" while it should be "Ask an administrator to archive its associated user first" (which is wrong, you are already an administrator, it was supposed to be "Switch company to archive its associated user first"). `check_access_rights` has ormcache, `check_access_rule` hasn't.

Reply:
After discussing with @mart-e it appears that finding the failing record(s) could be complicated. To be a bit optimized, we could have to use filter_access_rules, find the subset of failing records, and display heir name. It is a bit overkill, so I vote to keep Martin's solution: use only check_access_rights and keep a generic message in plural form ("You cannot archive contacts ..."). Keep it simple like they say.

Reply:
Hello :wave: 

I'm joining my colleagues to add a small remark.
Would be nice to add a:
`if self.env['res.users'].search_count([('partner_id', 'in', self.ids)])` before doing the check_access_rights so you avoid looping on partners and looking up users one by one. Since you only want to check "is there a single user linked to these partners?" the count should be fine.

And don't feel like we are bashing on you, this is actually a bit of a tricky piece of code so it takes many brains to come up with a good solution :smile: 

Aaaaand I'm gone! Good luck!

Reply:
Thanks for the detailed explanations, I learned a lot from this ;p I hope I understood correctly everyone's input, I guess we'll see once I push the next commit hehe